compiler/src/compiler/hir_to_mir.rs

```diff
@@ -0,0 +1,420 @@
+                body.push_with_new_id(
+                    id_generator,
+                    Expression::TraceCallEnds { return_value: call },
+                );
+            }
+            Expression::Reference(call)
+        }
+        hir::Expression::UseModule {
+            current_module,
+            relative_path,
+        } => Expression::UseModule {
+            current_module,
+            relative_path: mapping[&relative_path],
+            // The `UseModule` expression only exists in the generated `use`
+            // function. If a use fails, that's also the fault of the caller.
+            // Essentially, the `UseModule` expression works exactly like a
+            // `needs`.
+            responsible: responsible_for_needs,
+        },
+        hir::Expression::Needs { condition, reason } => {
+            let responsible =
+                body.push_with_new_id(id_generator, Expression::HirId(hir_id.clone()));
+            Expression::Call {
+                function: needs_function,
+                arguments: vec![mapping[&condition], mapping[&reason], responsible_for_needs],
+                responsible,
+            }
+        }
+        hir::Expression::Error { errors, .. } => {
+            let reason = body.push_with_new_id(
+                id_generator,
+                Expression::Text(if errors.len() == 1 {
+                    format!(
+                        "The code still contains an error: {}",
+                        errors.into_iter().next().unwrap().format_nicely(db)
+                    )
+                } else {
+                    format!(
+                        "The code still contains errors:\n{}",
+                        errors
+                            .into_iter()
+                            .map(|error| format!("- {}", error.format_nicely(db)))
+                            .join("\n"),
+                    )
+                }),
+            );
+            let responsible =
+                body.push_with_new_id(id_generator, Expression::HirId(hir_id.clone()));
+            Expression::Panic {
+                reason,
+                responsible,
+            }
+        }
+    };
+
+    let id = body.push_with_new_id(id_generator, expression);
+    mapping.insert(hir_id.clone(), id);
+
+    if config.trace_evaluated_expressions {
+        let hir_expression = body.push_with_new_id(id_generator, Expression::HirId(hir_id.clone()));
+        body.push_with_new_id(
+            id_generator,
+            Expression::TraceExpressionEvaluated {
+                hir_expression,
+                value: id,
+            },
+        );
+    }
+}
+
+impl CompilerError {
+    fn format_nicely(&self, db: &dyn HirToMir) -> String {
+        let (start_line, start_col) = db.offset_to_lsp(self.module.clone(), self.span.start);
+        let (end_line, end_col) = db.offset_to_lsp(self.module.clone(), self.span.end);
+
+        format!(
+            "{}, {}:{} – {}:{}: {}",
```

Comment:
A colon between file name and line/column is used by other programs as well and should also be clickable in VS Code

```suggestion
            "{}:{}:{} – {}:{}: {}",
```